File: DeviseAlgorithm/test_1/QuickSort/QuickSort.py
# 快速排序python实现
# 学号：20201050470
# 时间复杂度：O(nlog(n))
import random

def RandomArray(n):
    arr = []
    for i in range(n):
        arr.append(random.randint(-100000, 100000))
    return arr

# 生成线性同余序列
# def randlist(m, a, c, x0, n):
#     arr = [x0]
#     i: int
#     for i in range(n - 1):
#         arr.append((a * arr[i] + c) % m)
#     return arr


# 快速排序
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[0]
    left = [i for i in arr[1:] if i <= pivot]
    right = [i for i in arr[1:] if i > pivot]
    return quick_sort(left) + [pivot] + quick_sort(right)


# 测试算法
if __name__ == '__main__':
    n = (int)(input("随机数组长度："))
    # arr=randlist(100000, n, 0, 100, n)
    arr = RandomArray(n)
    print("排序前：\n",arr)
    print("\n排序后：\n",quick_sort(arr))

File: DeviseAlgorithm/test_1/QuickSort/QuickSort.java
/*
    快速排序Java实现
    学号：20201050470
    时间复杂度：O(nlog(n))
*/

import java.util.Scanner;

public class QuickSort {
    public static void main(String[] args) {
        QuickSortAlgorithm qs = new QuickSortAlgorithm();
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("请输入数组的长度：");
            int n = sc.nextInt();
            int [] a=qs.RandomArray(n,100000);   //生成随机数组
            System.out.println("随机序列");
            for (int j : a) {
                System.out.print(j+" ");
            }
            System.out.println();
            qs.quickSort(a,0,a.length-1);
            System.out.println("快速排序生成的非降序序列");
            for (int j : a) {
                System.out.print(j+" ");
            }
        }
    }
}
class QuickSortAlgorithm {
    int[]  RandomArray(int len,int max) {
        int[] arr = new int[len];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = (int) (Math.random() * max);
        }
         return arr;
 }
    //快速排序
    void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivot = partition(arr, low, high);
            quickSort(arr, low, pivot - 1);
            quickSort(arr, pivot + 1, high);
        }
    }
    //分区
    int partition(int[] arr, int low, int high) {
        int pivotkey = arr[low];
        while (low < high) {
            while (low < high && arr[high] >= pivotkey) {
                high--;
            }
            arr[low] = arr[high];
            while (low < high && arr[low] <= pivotkey) {
                low++;
            }
            arr[high] = arr[low];
        }
        arr[low] = pivotkey;
        return low;
    }
}